Archives/ClmFile.cpp

```diff
@@ -2,104 +2,69 @@
 	}
 
-	ClmFile::~ClmFile()
-	{
-		delete[] m_FileName;
-		delete[] m_IndexEntry;
-
-		if (m_FileHandle) {
-			CloseHandle(m_FileHandle);
-		}
-	}
+	ClmFile::~ClmFile() { }
 
 
 	// Reads in the header when the volume is first opened and does some
 	// basic error checking on the header.
-	// Returns nonzero if the header was read successfully and is intact
-	// Returns zero if an error occured
-	bool ClmFile::ReadHeader()
+	// Throws an error is problems encountered while reading the header.
+	void ClmFile::ReadHeader()
 	{
-		char buff[32];
-		unsigned long numBytes;
-		int i;
-
-		if (ReadFile(m_FileHandle, buff, 32, &numBytes, nullptr) == 0) return false;
-		if (memcmp(buff, "OP2 Clump File Version 1.0\x01A\0\0\0\0", 32)) return false;
+		clmFileReader->Read((char*)&clmHeader, sizeof(ClmHeader));
+		
+		if (!clmHeader.CheckFileVersion()) {
+			throw std::runtime_error("CLM file version is incorrect.");
+		}
 
-		// Read in the WaveFormatEx structure
-		if (ReadFile(m_FileHandle, &m_WaveFormat, sizeof(m_WaveFormat), &numBytes, nullptr) == 0) return false;
+		if (!clmHeader.CheckUnknown()) {
+			throw std::runtime_error("Unknown field in CLM header is incorrect.");
+		}
 
-		// Read remaining header info
-		if (ReadFile(m_FileHandle, m_Unknown, 6, &numBytes, nullptr) == 0) return false;
-		if (ReadFile(m_FileHandle, &m_NumberOfPackedFiles, 4, &numBytes, nullptr) == 0) return false;
-		if (m_NumberOfPackedFiles < 1) return false;
+		if (clmHeader.packedFilesCount < 1) {
+			throw std::runtime_error("Packed file count must be 1 or greater.");
+		}
 
-		// Allocate space
-		m_IndexEntry = new IndexEntry[m_NumberOfPackedFiles];
-		m_FileName = new char[m_NumberOfPackedFiles][9];
+		m_NumberOfPackedFiles = clmHeader.packedFilesCount;
 
-		// Read Index info
-		if (ReadFile(m_FileHandle, m_IndexEntry, m_NumberOfPackedFiles * sizeof(IndexEntry), &numBytes, nullptr) == 0)
-		{
-			delete[] m_FileName;
-			delete[] m_IndexEntry;
-			return false;
-		}
-		// Copy file names
-		for (i = 0; i < m_NumberOfPackedFiles; i++)
-		{
-			memcpy(m_FileName[i], &m_IndexEntry[i].fileName, 8);
-			m_FileName[i][8] = 0;	// NULL terminate it
+		indexEntries = std::vector<IndexEntry>(m_NumberOfPackedFiles);
+		clmFileReader->Read((char*)indexEntries.data(), m_NumberOfPackedFiles * sizeof(IndexEntry));
+		
+		// Null terminate all packed fileNames.
+		for (IndexEntry& indexEntry : indexEntries) {
+			indexEntry.fileName[8] = 0;
 		}
-
-		return true;
 	}
 
 
-
-	// Returns the internal file name of the internal file corresponding to index
-	const char* ClmFile::GetInternalFileName(int index)
+	// Returns the internal file name of the packed file corresponding to index.
+	// Throws an error if packed file index is not valid.
+	std::string ClmFile::GetInternalFileName(int index)
 	{
 		CheckPackedFileIndexBounds(index);
 
-		return m_FileName[index];
+		return indexEntries[index].GetFileName();
 	}
 
 	int ClmFile::GetInternalFileIndex(const char *internalFileName)
```

Comment:
I must be missing what you are implying. I thought the update was using a std::string by using IndexEntry::GetFileName?

Reply:
This one is misleading, since the comment is on a non-changed line of code. This is about `const char *internalFileName`, not about `.GetFileName()`. Sorry for the confusion.